Custom flat / model activation

BadLuckBurt opened the pull request:
These changes extend the custom model activation registration to include flats which was previously not possible. Any flat with a custom activation assigned will have a collider added to it's GameObject to allow for raycasting to succeed. Also, models with custom activations are excluded from the combinedMesh during dungeon assembly.

This allows modders to create custom activations for any model / flat they choose and enables them to implement things like resting at campfires, dungeon beds, etc. 

I have created a custom test dungeon (just a simple room) with a classic model (broomstick cabinet) and flat (campfire) as well as a replacement for both cases (a scarecrow flat is substituted for a white cube and the empty bookshelf is replaced with a bookshelf model I quickly made). Upon activation the cabinet, scarecrow cube and bookshelf will display a text message. Activating the campfire will present you with the Resting UI. 

You can find the dungeon here: [https://github.com/BadLuckBurt/dfu-custom-flat-model-activation-demo/releases](url) and it's source code here: [https://github.com/BadLuckBurt/dfu-custom-flat-model-activation-demo](url)

I have also tested to see what happens when an object already has an action assigned from classic DF using the throne room lever in PH and both actions fire after PlayerActivate.Update() finishes.

I'm open to suggestions and willing to make any changes required for this request to be accepted.

ajrb commented:
I reviewed this change before Burt submitted it and it is good as far as I am concerned. It does change the registration method name so will break mods that use this, but I am only aware of my Roleplay & Realism mod that this will affect and I will be releasing a new version of that once this is in a public build.

TheLacus commented on Assets/Scripts/Game/PlayerActivate.cs:
```diff
@@ -91,20 +91,42 @@ public static bool IsBuildingOpen(DFLocation.BuildingTypes buildingType)
+        private static Dictionary<string, CustomActivation> customActivations = new Dictionary<string, CustomActivation>();
 
-        public static bool RegisterModelActivation(uint modelID, ModelActivation modelActivation)
+        public static void RegisterCustomActivation(uint modelID, CustomActivation customActivation)
```
When i updated formula overrides i added a `Mod` parameter to [enforce load order](https://github.com/Interkarma/daggerfall-unity/blob/43e9f3cf4ed15d4a2079e65e03f7275d52e2be31/Assets/Scripts/Game/Formulas/FormulaHelper.cs#L2362) instead of always replacing/rejecting previously set delegates (also see [this issue](https://github.com/Interkarma/daggerfall-unity/issues/1668#issuecomment-589084075)). Might be worth to discuss if is also useful here.

BadLuckBurt replied:
I'm all for consistency so I will implement this and test before committing. 

If this feature becomes popular and mods start overriding each other, clear feedback will be important to know which mod registered what so I would like to add a Debug.Log during an overwrite if that's OK with everyone.

BadLuckBurt replied:
I have implemented this and first tests show it works as intended, any mod that is loaded after another will be allowed to overwrite any existing custom activation. I'll set up a few more tests where I have the mods hook in at different points of code execution.

TheLacus commented on Assets/Scripts/Game/PlayerActivate.cs:
```diff
@@ -91,20 +91,42 @@ public static bool IsBuildingOpen(DFLocation.BuildingTypes buildingType)
+            return CheckCustomActivation(goFlatName);
+        }
+
+        public static bool CheckCustomActivation(string goFlatModelName) {
```
I think is a bit confusing that this method is also public but has a different but similar name, especially because there is no xml summary. I personally would only use one name with three overloads.

BadLuckBurt replied:
Hazelnut raised the same point in a conversation we had but I decided to put it out as is and collect more feedback. 

I have renamed `CheckCustomActivation` to `HasCustomActivation` because looking at it again, you guys have a point.

TheLacus commented:
Custom activations are currently not [checked for distance](https://github.com/Interkarma/daggerfall-unity/blob/8860fb134220e0c8832c95761a6e4d640b89fe3b/Assets/Scripts/Game/PlayerActivate.cs#L665), aren't they? Maybe we should allow to provide a max distance (and store it with the delegate in a struct) or pass the distance for specific hit as a delegate parameter (or the hit itself; see [this](https://github.com/Interkarma/daggerfall-unity/blob/8860fb134220e0c8832c95761a6e4d640b89fe3b/Assets/Scripts/Game/PlayerActivate.cs#L41)). [These constants](https://github.com/Interkarma/daggerfall-unity/blob/8860fb134220e0c8832c95761a6e4d640b89fe3b/Assets/Scripts/Game/PlayerActivate.cs#L68) should also be accessible i think.

ajrb commented on Assets/Scripts/Game/PlayerActivate.cs:
```diff
@@ -91,20 +91,42 @@ public static bool IsBuildingOpen(DFLocation.BuildingTypes buildingType)
-        // Allow mods to register custom model activation methods.
-        public delegate void ModelActivation(Transform transform);
-        private static Dictionary<string, ModelActivation> customModelActivations = new Dictionary<string, ModelActivation>();
+        // Allow mods to register custom flat / model activation methods.
+        public delegate void CustomActivation(Transform transform);
+        private static Dictionary<string, CustomActivation> customActivations = new Dictionary<string, CustomActivation>();
```
What do you mean? Sorry I don't get what you're suggesting here.

BadLuckBurt replied:
I'm not too familiar with C# yet so I read up on the `readonly` keyword and I understand why that would be a good thing so I have added that but to implement the `Action<Transform>` I'd need a little more to go on. I looked them up but I don't see any benefit compared to the way things are now, probably due to my inexperience.

What are you aiming for and what would be the benefit?

TheLacus replied:
I was suggesting that a named delegate doesn't add much value in this case, but is not really important, especially now that we are considering adding a new `Hit` argument. I'm happy with how things are now :)

ajrb replied:
I'm still curious to understand what your suggestion was... I just have no idea what making the delegate type use Action<Template> even meant. :-)

TheLacus replied:
_Transform_, not _Template_. See [Action{T}](https://docs.microsoft.com/en-us/dotnet/api/system.action-1?view=netframework-4.8).

ajrb replied:
Sorry, mixed up the words. I was not aware of the `Action<T>` stuff before.

ajrb commented:
> Custom activations are currently not [checked for distance](https://github.com/Interkarma/daggerfall-unity/blob/8860fb134220e0c8832c95761a6e4d640b89fe3b/Assets/Scripts/Game/PlayerActivate.cs#L665), aren't they? Maybe we should allow to provide a max distance (and store it with the delegate in a struct) or pass the distance for specific hit as a delegate parameter (or the hit itself; see [this](https://github.com/Interkarma/daggerfall-unity/blob/8860fb134220e0c8832c95761a6e4d640b89fe3b/Assets/Scripts/Game/PlayerActivate.cs#L41)). [These constants](https://github.com/Interkarma/daggerfall-unity/blob/8860fb134220e0c8832c95761a6e4d640b89fe3b/Assets/Scripts/Game/PlayerActivate.cs#L68) should also be accessible i think.

Good idea, I think passing the hit same as your interface is best - let the implementation do whatever. For some reason I thought that the distance had already been checked in PlayerActivate. And making those constants available for this too.

Interkarma commented:
Thank you for help with review @ajrb and @TheLacus, I very much appreciate it. :)

I'm not as involved with the modding side of things as you two, so I'll wait for both of your approvals before merging.

BadLuckBurt commented:
> Custom activations are currently not [checked for distance](https://github.com/Interkarma/daggerfall-unity/blob/8860fb134220e0c8832c95761a6e4d640b89fe3b/Assets/Scripts/Game/PlayerActivate.cs#L665), aren't they? Maybe we should allow to provide a max distance (and store it with the delegate in a struct) or pass the distance for specific hit as a delegate parameter (or the hit itself; see [this](https://github.com/Interkarma/daggerfall-unity/blob/8860fb134220e0c8832c95761a6e4d640b89fe3b/Assets/Scripts/Game/PlayerActivate.cs#L41)). [These constants](https://github.com/Interkarma/daggerfall-unity/blob/8860fb134220e0c8832c95761a6e4d640b89fe3b/Assets/Scripts/Game/PlayerActivate.cs#L68) should also be accessible i think.

I gave this some thought and I'd propose to allow the mod to provide a max distance and store it in the struct like you suggested. If no distance is provided it should probably be set to DefaultActivationDistance. While implementing the load priority stuff I've already made a struct so it's no issue to add a distance property to it.

When you say 'these constants should also be accessible', do you mean you want me to make them public? No problem if that's the case, I just want to make sure before I do.

BadLuckBurt commented:
I believe I have satisfied the requests made, I've summarised them below. Apologies in case I managed to miss anything, please point it out and I'll get to it ASAP.

**1. Added struct CustomModActivation to store custom activations using the following properties:**
- Action = the custom activation method to be called
- ActivationDistance = the distance in Unity units from which the object can be activated
- Provider = the mod that registered the custom activation

**2. Changed dictionary name to customModActivations**

**3. Method naming**
- RegisterCustomActivation for both model and flat entry points
- HandleRegisterCustomActivation for the actual registration

**4. Both RegisterCustomActivation methods have new parameters**
- provider (Mod instance that is registering the custom activation)
- activationDistance (This is the activation distance in Unity units, if none is supplied it will default to the DefaultActivationDistance const)

**5. The HandleRegisterCustomActivation method now checks for existing custom activations on a goFlatModelName and if found, compare load order priorities to determine wether an override is allowed**

**6. CheckCustomActivation method has been renamed to HasCustomActivation**

**7. PlayerActivate.Update now checks the activation distance and when calling the custom activation's action, will supply the RaycastHit as a parameter instead of the hit.transform**

**8. Made the activation distances in PlayerActivate.cs public so they can be accessed from outside the class**

BadLuckBurt commented:
I merged Interkarma's current master with my fork's master and updated this branch as well to ensure the commits will not interfere with anything.

ajrb commented:
Looks great Burt, good work.

@Interkarma This is ready to merge from my POV.

Interkarma commented:
Looks good from here too. Thanks for the changes @BadLuckBurt. Will merge this one now. And I really appreciate the help @ajrb and @TheLacus. Cheers guys! :)

BadLuckBurt commented:
> 
> 
> Looks good from here too. Thanks for the changes @BadLuckBurt. Will merge this one now. And I really appreciate the help @ajrb and @TheLacus. Cheers guys! :)

Thank you for the merge. I will be making a thread on the forums on how to use this later today. Cheers!
